### PR TITLE
Feature/add near

### PR DESCRIPTION
**Pull request by WhiteNik16**

### CODE REVIEW

File: packages/provider/src/helpers/near.ts [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+    const transaction = await provider.txStatus(txhash, 'unnused')
+    return providers.getTransactionLastResult(transaction)
+  }
+}
+
+export const nearProviderBase = new _NearProviderBase({
+  // createAccessKeyFor: 'example-account.testnet',
+})
+
+export const convertNearAmountToYocto = (amount: string) => {
```
Comment (napalmpapalam) [inline packages/provider/src/helpers/near.ts:121]: 1. add return type
2. rename to `nearToYocto`

File: packages/provider/src/helpers/near.ts [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+}
+
+export function getNearExplorerTxUrl(
+  explorerUrl: string | Chain,
+  txHash: string,
+) {
+  return `${explorerUrl}/transactions/${txHash}`
+}
+
+export function getNearExplorerAddressUrl(
```
Comment (napalmpapalam) [inline packages/provider/src/helpers/near.ts:140]: add return types

File: packages/provider/src/providers/near.ts [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+
+export class NearProvider extends BaseNear implements ProviderProxy {
+  constructor(provider: RawProvider) {
+    super(provider)
+  }
+
+  static get providerType(): Providers {
+    return Providers.Near
+  }
+}
```
Comment (napalmpapalam) [inline packages/provider/src/providers/near.ts:13]: no need to do it with near provider coz u will have only one implementation

File: packages/provider/src/providers/base-near.ts [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+  #isConnected = false
+  #chainId?: ChainId
+  #address?: string
+
+  constructor(provider: RawProvider) {
+    super()
+    this.#web3 = new window.Web3(provider as unknown as HttpProvider)
+    this.#provider = (<unknown>this.#web3?.currentProvider) as NearProviderType
+    this.#isConnected = false
```
Comment (napalmpapalam) [inline packages/provider/src/providers/base-near.ts:38]: no need to make additional `#isConnected` field, in most cases `isConnected` is `Boolean(this.#chainId && this.#address)`

File: packages/provider/src/providers/base-near.ts [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+  readonly #web3: Web3
+
+  #isConnected = false
+  #chainId?: ChainId
+  #address?: string
+
+  constructor(provider: RawProvider) {
+    super()
+    this.#web3 = new window.Web3(provider as unknown as HttpProvider)
+    this.#provider = (<unknown>this.#web3?.currentProvider) as NearProviderType
```
Comment (napalmpapalam) [inline packages/provider/src/providers/base-near.ts:37]: a u sure that this is needed ?

File: packages/provider/src/providers/base-near.ts [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+  }
+
+  async init(): Promise<void> {
+    try {
+      await this.#provider.init()
+      this._updateProviderState()
+    } catch (error) {
+      handleNearError(error as NearProviderRpcError)
+    }
+  }
```
Comment (napalmpapalam) [inline packages/provider/src/providers/base-near.ts:60]: emit initiated event

File: packages/provider/src/providers/base-near.ts [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+
+  _updateProviderState(): void {
+    const networkId = this.#provider.selector?.options.network.networkId
+
+    this.#address = this.#provider?.accountId || ''
+    this.#chainId = networkId || NEAR_CHAINS.mainnet
+  }
```
Comment (napalmpapalam) [inline packages/provider/src/providers/base-near.ts:67]: 1. `#updateProviderState`
2. made default chain id devnet

File: packages/provider/src/providers/base-near.ts [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+
+  async switchChain(chainId: ChainId): Promise<void> {
+    this.#chainId = chainId
```
Comment (napalmpapalam) [inline packages/provider/src/providers/base-near.ts:70]: emit event

File: packages/provider/src/providers/base-near.ts [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+  }
+
+  async connect(): Promise<void> {
+    try {
+      await this.#provider.signIn()
```
Comment (napalmpapalam) [inline packages/provider/src/providers/base-near.ts:75]: emit event

File: packages/provider/src/providers/base-near.ts [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+    } catch (error) {
+      handleNearError(error as NearProviderRpcError)
+    }
+  }
+
+  async disconnect(): Promise<void> {
+    try {
+      await this.#provider.signOut()
+      this._updateProviderState()
```
Comment (napalmpapalam) [inline packages/provider/src/providers/base-near.ts:84]: emit event

File: packages/provider/src/providers/base-near.ts [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+      )
+      return txReceipt.transactionHash
+    } catch (error) {
+      handleEthError(error as EthProviderRpcError)
+    }
+
+    return ''
+  }
+
+  async signAndSendTransaction(txRequestBody: TxRequestBody) {
```
Comment (napalmpapalam) [inline packages/provider/src/providers/base-near.ts:117]: add return types

File: packages/provider/src/providers/base-near.ts [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+    return getNearExplorerTxUrl(chain, txHash)
+  }
+
+  getAddressUrl(chain: Chain, address: string) {
+    return getNearExplorerAddressUrl(chain, address)
+  }
+
+  async signAndSendTx(tx: TxRequestBody): Promise<TransactionResponse> {
+    try {
+      const txReceipt = await this.#web3.eth.sendTransaction(
```
Comment (napalmpapalam) [inline packages/provider/src/providers/base-near.ts:106]: eth? it's not work in that way

File: packages/provider/src/helpers/near.ts [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+    const provider = new providers.JsonRpcProvider({ url: network.nodeUrl })
+
+    // Retrieve transaction result from the network
+    const transaction = await provider.txStatus(txhash, 'unnused')
+    return providers.getTransactionLastResult(transaction)
+  }
+}
+
+export const nearProviderBase = new _NearProviderBase({
+  // createAccessKeyFor: 'example-account.testnet',
```
Comment (napalmpapalam) [inline packages/provider/src/helpers/near.ts:118]: why it's commented?

File: packages/provider/src/types/near.ts [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+
+export type NearTxRequestBody = {
+  contractId: string
+  method: string
+  args?: Record<string, unknown>
+  gas?: string
+  deposit?: string
+}
+
+export type NearProviderType = typeof nearProviderBase
```
Comment (napalmpapalam) [inline packages/provider/src/types/near.ts:15]: `InstanceType<NearRawProvider>`

File: packages/provider/src/web3.ts [hunk 15]
```diff
@@ -62,9 +62,11 @@ export class Web3 {
       : undefined
     const phantomProvider = window?.solana
     const solflareProvider = window?.solflare
+    const nearProvider = [nearProviderBase]
```
Comment (napalmpapalam) [inline packages/provider/src/web3.ts:65]: create instance here

File: yarn.lock [hunk 16]
```diff
@@ -10807,6 +10950,13 @@ __metadata:
...
+"u3@npm:^0.1.1":
+  version: 0.1.1
+  resolution: "u3@npm:0.1.1"
+  checksum: d55f396c607b0a2340d6165526b5143a29b369e7e0397b04f79633db77cd668f2da713feb7adb4f93a588e82a388ff995d9a2b16123d0cb02fe394fd2f26b529
+  languageName: node
+  linkType: hard
+
 "ultron@npm:~1.1.0":
   version: 1.1.1
   resolution: "ultron@npm:1.1.1"
```
Comment (napalmpapalam) [inline yarn.lock:10962]: change log

**Review by napalmpapalam — CHANGES_REQUESTED** [review]

File: .gitattributes [hunk 17]
```diff
@@ -1,13 +1,3 @@
...
-# Declare files that will always have LF line endings on checkout.
-*.* text eol=lf
-
-# Denote all files that are truly binary and should not be modified.
-*.png binary
-*.jpg binary
-*.otf binary
-*.ttf binary
-*.woff binary
-*.woff2 binary
```
[the remark this answers is not in the document]
Comment (WhiteNik16) [reply .gitattributes:13]: fixed

File: .gitattributes [hunk 18]
```diff
@@ -1,13 +1,3 @@
...
+/.yarn/releases/** binary
+/.yarn/plugins/** binary
+*.zip binary
```
[the remark this answers is not in the document]
Comment (WhiteNik16) [reply .gitattributes:3]: fixed

File: packages/provider/src/helpers/near.ts [hunk 19]
```diff
@@ -0,0 +1,187 @@
...
+  Chain,
+  ENearWalletId,
+  NearProviderRpcError,
+  NearTxRequestBody,
+} from "../types";
+
+const THIRTY_TGAS = '300000000000000'
+const NO_DEPOSIT = '0'
+
+export class _NearProviderBase {
```
[the remark this answers is not in the document]
Comment (WhiteNik16) [reply packages/provider/src/helpers/near.ts:27]: fixed

File: packages/provider/src/helpers/near.ts [hunk 20]
```diff
@@ -0,0 +1,187 @@
...
+} from '../enums'
+import { errors } from '../errors'
+import {
+  Chain,
+  ENearWalletId,
+  NearProviderRpcError,
+  NearTxRequestBody,
+} from "../types";
+
+const THIRTY_TGAS = '300000000000000'
```
[the remark this answers is not in the document]
Comment (WhiteNik16) [reply packages/provider/src/helpers/near.ts:24]: fixed

File: packages/provider/src/helpers/near.ts (hunk 1 above)
Comment (WhiteNik16) [reply packages/provider/src/helpers/near.ts:121]: fixed

File: packages/provider/src/helpers/near.ts [hunk 21]
```diff
@@ -0,0 +1,187 @@
...
+  // createAccessKeyFor: 'example-account.testnet',
+})
+
+export const convertNearAmountToYocto = (amount: string) => {
+  // converts NEAR amount into yoctoNEAR (10^-24)
+
+  return utils.format.parseNearAmount(amount)
+}
+
+export const convertYoctoToNearAmount = (amount: string) => {
```
[the remark this answers is not in the document]
Comment (WhiteNik16) [reply packages/provider/src/helpers/near.ts:127]: fixed

File: packages/provider/src/helpers/near.ts (hunk 2 above)
Comment (WhiteNik16) [reply packages/provider/src/helpers/near.ts:140]: fixed

File: packages/provider/src/providers/near.ts (hunk 3 above)
Comment (WhiteNik16) [reply packages/provider/src/providers/near.ts:13]: fixed

File: packages/provider/src/providers/base-near.ts (hunk 5 above)
Comment (WhiteNik16) [reply packages/provider/src/providers/base-near.ts:37]: fixed

File: packages/provider/src/providers/base-near.ts (hunk 6 above)
Comment (WhiteNik16) [reply packages/provider/src/providers/base-near.ts:60]: fixed

File: packages/provider/src/providers/base-near.ts (hunk 7 above)
Comment (WhiteNik16) [reply packages/provider/src/providers/base-near.ts:67]: fixed

File: packages/provider/src/providers/base-near.ts (hunk 8 above)
Comment (WhiteNik16) [reply packages/provider/src/providers/base-near.ts:70]: fixed

File: packages/provider/src/providers/base-near.ts (hunk 9 above)
Comment (WhiteNik16) [reply packages/provider/src/providers/base-near.ts:75]: fixed

File: packages/provider/src/providers/base-near.ts (hunk 11 above)
Comment (WhiteNik16) [reply packages/provider/src/providers/base-near.ts:117]: fixed

File: packages/provider/src/providers/base-near.ts (hunk 12 above)
Comment (WhiteNik16) [reply packages/provider/src/providers/base-near.ts:106]: fixed

File: packages/provider/src/helpers/near.ts (hunk 13 above)
Comment (WhiteNik16) [reply packages/provider/src/helpers/near.ts:118]: fixed

File: packages/provider/src/web3.ts (hunk 15 above)
Comment (WhiteNik16) [reply packages/provider/src/web3.ts:65]: fixed

File: packages/provider/src/providers/base-near.ts (hunk 5 above)
Comment (WhiteNik16) [reply packages/provider/src/providers/base-near.ts:37]: Yes

File: packages/provider/src/types/near.ts (hunk 14 above)
Comment (WhiteNik16) [reply packages/provider/src/types/near.ts]: fixed

File: packages/provider/src/enums/chains.ts [hunk 22]
```diff
@@ -8,3 +8,8 @@ export enum SolanaChains {
   testnet = 'testnet',
   mainnet = 'mainnet',
 }
+
+export enum NEAR_CHAINS {
```
Comment (napalmpapalam) [inline packages/provider/src/enums/chains.ts:12]: pascal case

File: packages/provider/src/enums/chains.ts [hunk 23]
```diff
@@ -8,3 +8,8 @@ export enum SolanaChains {
...
+  testnet = 'testnet',
+  mainnet = 'mainnet',
```
Comment (napalmpapalam) [inline packages/provider/src/enums/chains.ts:14]: refactor SolanaChains and NEAR_CHAINS keys to Pascal case

**Review by napalmpapalam — CHANGES_REQUESTED** [review]

**Pull request merged**